test/cli_test.go

```diff
@@ -0,0 +1,379 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"testing"
+	"text/template"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/GoogleCloudPlatform/terraform-validator/converters/google"
+	"github.com/GoogleCloudPlatform/terraform-validator/version"
+)
+
+const (
+	samplePolicyPath = "../testdata/sample_policies"
+	defaultAncestry  = "organization/12345/folder/67890"
+	defaultProject   = "foobar"
+)
+
+var (
+	data      *testData
+	tfvBinary string
+)
+
+// testData represents the full dataset that is used for templating terraform
+// configs. It contains Google API resources that are expected to be returned
+// after converting the terraform plan.
+type testData struct {
+	// is not nil - Terraform 12 version used
+	TFVersion string
+	// provider "google"
+	Provider map[string]string
+	Project  map[string]string
+	Ancestry string
+}
+
+// init initializes the variables used for testing. As tests rely on
+// environment variables, the parsing of those are only done once.
+func init() {
+	cwd, err := os.Getwd()
+	if err != nil {
+		log.Fatalf("cannot get current directory: %v", err)
+	}
+	tfvBinary = filepath.Join(cwd, "..", "bin", "terraform-validator")
+	project, ok := os.LookupEnv("TEST_PROJECT")
+	if !ok {
+		log.Printf("Missing required env var TEST_PROJECT. Default (%s) will be used.", defaultProject)
+		project = defaultProject
+	}
+	credentials, ok := os.LookupEnv("TEST_CREDENTIALS")
+	if ok {
+		// Make credentials path relative to repo root rather than
+		// test/ dir if it is a relative path.
+		if !filepath.IsAbs(credentials) {
+			credentials = filepath.Join(cwd, "..", credentials)
+		}
+	} else {
+		log.Printf("missing env var TEST_CREDENTIALS, will try to use Application Default Credentials")
+	}
+	ancestry, ok := os.LookupEnv("TEST_ANCESTRY")
+	if !ok {
+		log.Printf("Missing required env var TEST_ANCESTRY. Default (%s) will be used.", defaultAncestry)
+		ancestry = defaultAncestry
+	}
+	providerVersion := "1.20"
+	if version.TF12 == version.LeastSupportedVersion() {
+		providerVersion = "2.12.0"
+	}
+	data = &testData{
+		TFVersion: version.LeastSupportedVersion(),
+		Provider: map[string]string{
+			"version":     providerVersion,
+			"project":     project,
+			"credentials": credentials,
+		},
+		Project: map[string]string{
+			"Name":               "My Project Name",
+			"ProjectId":          "my-project-id",
+			"BillingAccountName": "012345-567890-ABCDEF",
+		},
+		Ancestry: ancestry,
+	}
+}
+
+// TestCLI tests the "convert" and "validate" subcommand against a generated .tfplan file.
+func TestCLI(t *testing.T) {
+	// Define the reusable rules to be use for the test cases.
+	type rule struct {
```

Comment:
Let's keep naming consistent:

```suggestion
	type constraint struct {
```

Reply:
yes. Fixed.